lib/Sema/SemaExpr.cpp

```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
       << ResultType;
     return ExprError();
   }
-  return new (Context) CountBoundsExpr(Kind, PromotedCountExpr, BoundsKWLoc,
+  return new (Context) CountBoundsExpr(Kind, CountExpr, BoundsKWLoc,
                                        RParenLoc);
 }
 
+/// \brief Validate the type of an argument expression to a bounds
+/// expression. If valid, return the unqualified pointee type.  Otherwise
+/// return null.
+const Type *Sema::ValidateBoundsExprArgument(Expr *Arg) {
+  QualType ArgType = Arg->getType();
+  const Type *ArgPointerType = ArgType->getAs<PointerType>();
+  if (!ArgPointerType) {
+    Diag(Arg->getLocStart(),
+         diag::err_typecheck_pointer_type_expected) << ArgType;
+    return nullptr;
+  }
+  QualType ArgTypePointee = ArgPointerType->getPointeeType();
+  // For the Checked C extension, we don't expect to see of the clang
+  // extensions for qualifiers.  Assert if we do.
```

Comment:
They do support them, but they are treated as attributes on types, not modifiers.    I'll add a longer comment explaining what is going on with clang extensions to type qualifiers.